docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+
+#### Story 7
+
+As a service provider I want to be able to offer Managed Kubernetes clusters by using CAPI referencing my own managed control plane implementation that satisfies Cluster API contracts.
+
+### Design
+
+Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.
+
+#### Core Cluster API changes
+
+This proposal does not introduce any breaking changes for the existing "core" API. More specifically:
+
+The existing Cluster API types are already able to omit the `<Infra>Cluster`:
+
+- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
+- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.
+
+In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  hostname: 'name-1234567890.region.elb.amazonaws.com'
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  ip: 10.40.85.102
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+This type will be the primary way for both `<Infra>Cluster` or `ControlPlane Provider` to communicate the ClusterAddress to "core" Cluster API controllers.
+
+The `<Infra>Cluster` object which is currently using the `spec.controlPlaneEndpoint` for the same scope will continue to work because "core" Cluster API controllers will continue to recognize when this field is set and take care of generating the `ClusterEndpoint` automatically; however this mechanism should be considered as a temporary machinery to migrate to the new CRD, and it will be removed in future versions of Cluster API. In addition, once the legacy behavior is removed, we will deprecate and eventually remove the `spec.controlPlaneEndpoint` field from the `Cluster` CustomResourceDefinition, and recommend that providers do the same for their `<Infra>Cluster` CustomResourceDefinitions as well.
+
+Notes:
+
+- The `hostname` and `ip` fields are mutually exclusive
+- The `type` field has been introduced to allow to extend the usage of this CRD to address https://github.com/kubernetes-sigs/cluster-api/issues/5295 in a future iterations
```

Comment:
I'm not sure I'm seeing how introducing a "type" now enables the use case any better than not introducing a type at all for now. I think we'll be in a better position to reason about the best way to add it when we actually need to disambiguate between types.
By introducing it now I think we'll add friction for existing use cases which might implement a private topology, in which case forcing them to have an endpoint of type ExternalControlPlaneEndpoint seems inappropriate.

Reply:
We can add it later probably, and default it. Adding it now might make it explicit, and less to deal with conversion later

Reply:
If we don't add it now (and a type=ControlPlaneEndpoint is implied), then adding it later should allow us to easily make that default assignment (if type=="" then set type=ControlPlaneEndpoint) in future versions for back-compat.

Do we have consensus that such an approach (wait to introduce until we have a > 1 type use case) is preferable to explicitly including a new `type` property in the scope of this PR?

Reply:
I'm fine with adding it later until we have a better idea about other values

Reply:
Later sounds good

Reply:
Done